types/ramda/index.d.ts

```diff
@@ -1469,55 +1450,11 @@ export function pickBy<T>(pred: ObjPred<T>): <U, V extends T>(obj: V) => U;
-export function pipe<V0, V1, T1, T2, T3, T4, T5, T6, T7, T8>(fn0: (x0: V0, x1: V1) => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8): (x0: V0, x1: V1) => T8;
-export function pipe<V0, V1, V2, T1, T2, T3, T4, T5, T6, T7, T8>(fn0: (x0: V0, x1: V1, x2: V2) => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8): (x0: V0, x1: V1, x2: V2) => T8;
-
-export function pipe<T1, T2, T3, T4, T5, T6, T7, T8, T9>(fn0: () => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8, fn8: (x: T8) => T9): () => T9;
-export function pipe<V0, T1, T2, T3, T4, T5, T6, T7, T8, T9>(fn0: (x0: V0) => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8, fn8: (x: T8) => T9): (x0: V0) => T9;
-export function pipe<V0, V1, T1, T2, T3, T4, T5, T6, T7, T8, T9>(fn0: (x0: V0, x1: V1) => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8, fn8: (x: T8) => T9): (x0: V0, x1: V1) => T9;
-export function pipe<V0, V1, V2, T1, T2, T3, T4, T5, T6, T7, T8, T9>(fn0: (x0: V0, x1: V1, x2: V2) => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8, fn8: (x: T8) => T9): (x0: V0, x1: V1, x2: V2) => T9;
-
-export function pipe<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>(fn0: () => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8, fn8: (x: T8) => T9, fn9: (x: T9) => T10): () => T10;
-export function pipe<V0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>(fn0: (x0: V0) => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8, fn8: (x: T8) => T9, fn9: (x: T9) => T10): (x0: V0) => T10;
-export function pipe<V0, V1, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>(fn0: (x0: V0, x1: V1) => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8, fn8: (x: T8) => T9, fn9: (x: T9) => T10): (x0: V0, x1: V1) => T10;
-export function pipe<V0, V1, V2, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>(fn0: (x0: V0, x1: V1, x2: V2) => T1, fn1: (x: T1) => T2, fn2: (x: T2) => T3, fn3: (x: T3) => T4, fn4: (x: T4) => T5, fn5: (x: T5) => T6, fn6: (x: T6) => T7, fn7: (x: T7) => T8, fn8: (x: T8) => T9, fn9: (x: T9) => T10): (x0: V0, x1: V1, x2: V2) => T10;
+export function pipe<
+ TFunctions extends AnyFunction[],
```

Comment:
> 

It cannot.

Reply:
> how can Typescript constraint the intermedia functions type

Unfortunately I don't believe it can, using this implementation. No inferring s, and n2 expects a number but is given a string:
![image](https://user-images.githubusercontent.com/7280931/143733795-09b14d80-f065-448c-ac0a-de23b9bf5c09.png)

Reply:
Yeah ... that one is tricky 🤔 Any idea if we can add some extra magical things?

Reply:
I did some ramda optimizes recently, maybe you can optimize `compose` and `pipe` based on my pr:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57393/files

Reply:
@yordis We recently merged a PR from @adispring that included an optimised compose and pipe, you can check it out here: #57393

Reply:
@Nicholaiii does that mean it would deprecate this one or should I rebase?

Reply:
Other things broke,

I made it to work outside ramda and I was sure that it did work properly, but now I can't tell if it is failing because of my test cases or other parts of the typespecs.

I am feeling a bit discouraged 😞

Reply:
> I am feeling a bit discouraged 😞
Yeah, I get that :\ 

I don't know, I like the idea but if it doesn't work it isn't feasible to merge.

Reply:
I will continue figuring out the issues with other functions specs

Reply:
The thing i like in current `pipe` implementation is that Typescipt can match return and arguments of two consecutive functions, because it gives a lot of type safety. Unfortunately we can do this in Typescript only by declaring ten overloads of the same function. 

In implementation with `AnyFunction` this case will be considered valid:
```ts
function a(value: Apple): Banana {};

function b(value: Pear): Orange {};

const c: (value: Apple) => Orange = pipe(a, b) // <- types Banana and Pear are incompatible
```

but error will be given in current implementation.

Reply:
@NEWESTERS
Yes, I agree, the current implemention, especially with the work from @adispring is really strong, and gives type-safety up to 7 functions. If we could achieve the same with less complexity it would be great, but as is rn it's nice.

Reply:
I think we can find a way to concise the current `pipe` implementation.